validation/validate.py

```diff
@@ -0,0 +1,215 @@
+        _type = template_keyword['type']
+        if not type_matches(input_dict[keyword], _type):
+            raise InputError("incorrect type for keyword: {0}, expected '{1}' type".format(keyword, _type))
+
+    # fill missing input keywords with template defaults
+    for keyword in set(template_keywords).difference(set(input_keywords)):
+        template_keyword = list(filter(lambda x: x['keyword'] == keyword, template_dict['keywords']))[0]
+        _default = template_keyword['default']
+        input_dict[keyword] = _default
+
+    # if this node contains sections, we descend into these and verify these in turn
+    for section in template_sections:
+        input_section = input_dict[section]
+        template_section = list(filter(lambda x: x['section'] == section, template_dict['sections']))[0]
+        input_dict[section] = validate_node(input_section, template_section)
+
+    return input_dict
+
+
+def check_predicates_node(input_dict, input_dict_node, template_dict_node):
+    '''
+    Check keyword predicates in a node. A node is either the root of the input or it is a section.
+    The node can contain keywords or other sections which we traverse recursively.
+
+    This function does not modify input arguments.
+
+    Can raise either InputError or TemplateError.
+    '''
+    input_sections = [x for x in input_dict_node if type(input_dict_node[x]).__name__ == 'dict']
+    input_keywords = list(filter(lambda x: x not in input_sections, input_dict_node.keys()))
+
+    if 'sections' in template_dict_node:
+        template_sections = [x['section'] for x in template_dict_node['sections']]
+    else:
+        template_sections = []
+
+    if 'keywords' in template_dict_node:
+        template_keywords = [x['keyword'] for x in template_dict_node['keywords']]
+    else:
+        template_keywords = []
+
+    # go through the predicates
+    for keyword in input_keywords:
+        template_keyword = list(filter(lambda x: x['keyword'] == keyword, template_dict_node['keywords']))[0]
+        value = input_dict_node[keyword]
+        if 'predicates' in template_keyword:
+            for predicate in template_keyword['predicates']:
+                try:
+                    r = eval(predicate)
+                except SyntaxError:
+                    raise TemplateError('syntax error in predicate {} in keyword {}'.format(predicate, keyword))
+                if not r:
+                    raise InputError('predicate {} failed in keyword {}'.format(predicate, keyword))
+
+    # if this node contains sections, we descend into these and verify these in turn
+    for section in template_sections:
+        input_section = input_dict_node[section]
+        template_section = list(filter(lambda x: x['section'] == section, template_dict_node['sections']))[0]
+        check_predicates_node(input_dict, input_section, template_section)
+
+
+if __name__ == '__main__':
+    input_dict = read_yaml_file('example.yml')
```

Comment:
When this consolidates with code and tests in the right directories, we'll need to add tests for directly passing a `dict` inside and using a JSON file.

Reply:
Agreed - this is the plan.

Reply:
But I still might start with file-based tests until the concrete solidifies a bit. The reason is that it is easier to change files rather than change/edit nested dictionaries.

Reply:
The testing with `dict`-s can come later. I have to investigate how to automate it/randomize it with Hypothesis and then we can have a one (or few) functions generating test cases as `dict`-s with serialization to JSON/YAML done separately.

Reply:
Actually I now realized that I will need some dicts for the tests. I will add a number of hand-written tests, later we can make it more sophisticated with hypothesis.

Reply:
Can I resolve this discussion or do you want to follow up with an issue?

Reply:
No need to follow up with an issue.